pumps/sql_aggregate.go

```diff
@@ -95,31 +95,58 @@ func (c *SQLAggregatePump) Init(conf interface{}) error {
 }
 
 func (c *SQLAggregatePump) WriteData(ctx context.Context, data []interface{}) error {
-	c.log.Debug("Attempting to write ", len(data), " records...")
-
-	analyticsPerOrg := analytics.AggregateData(data, c.SQLConf.TrackAllPaths, c.SQLConf.IgnoreTagPrefixList, c.SQLConf.StoreAnalyticsPerMinute)
+	dataLen := len(data)
+	c.log.Debug("Attempting to write ", dataLen, " records...")
+
+	startIndex := 0
+	endIndex := dataLen
+	for i := 0; i < dataLen; i++ {
+		if c.SQLConf.TableSharding {
+			recDate := data[startIndex].(analytics.AnalyticsRecord).TimeStamp.Format("20060102")
+			nextRecDate := data[i].(analytics.AnalyticsRecord).TimeStamp.Format("20060102")
+
+			if i != dataLen-1 && recDate == nextRecDate { // write records belong to same day at once
+				continue
+			}
 
-	for orgID, ag := range analyticsPerOrg {
-		// ag.DiscardAggregations([]string{"keyendpoints", "oauthendpoints", "apiendpoints"})
+			endIndex = i
+			if endIndex == dataLen-1 {
+				endIndex = dataLen
+			}
 
-		for _, d := range ag.Dimensions() {
-			rec := analytics.SQLAnalyticsRecordAggregate{
-				OrgID:          orgID,
-				TimeStamp:      ag.TimeStamp.Unix(),
-				Counter:        *d.Counter,
-				Dimension:      d.Name,
-				DimensionValue: d.Value,
+			table := "tyk_aggregated_" + recDate
+			c.db = c.db.Table(table)
+			if !c.db.Migrator().HasTable(table) {
+				c.db.AutoMigrate(&analytics.SQLAnalyticsRecordAggregate{})
 			}
-			rec.ProcessStatusCodes()
+		} else {
+			i = dataLen // write all records at once for non-sharded case, stop for loop after 1 iteration
+		}
 
-			resp := c.db.WithContext(ctx).Create(rec)
-			if resp.Error != nil {
-				return resp.Error
+		analyticsPerOrg := analytics.AggregateData(data[startIndex:endIndex], c.SQLConf.TrackAllPaths, c.SQLConf.IgnoreTagPrefixList, c.SQLConf.StoreAnalyticsPerMinute)
```

Comment:
That's smart approach to deal with multi-hour records. 
But it took me some time to understand how it works.
Maybe makes sense to extend code comments a bit.

Reply:
added.